### PR TITLE
Fix gcs actor manager destroy actor crash bug

### PR DESCRIPTION
**Pull request by ffbin**
## Why are these changes needed?
Problem:
If an actor is deleted immediately after the user creates the actor, reference counter may return a reply to the request of WaitForActorOutOfScope to GCS server, and GCS server will destroy the actor. The actor creation is asynchronous, it may be destroyed before the actor creation is completed.
Solution:
We will check whether the created actors are deleted.

## Related issue number
https://github.com/ray-project/ray/issues/9267

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: src/ray/gcs/gcs_server/test/gcs_actor_manager_test.cc [hunk 1]
```diff
@@ -576,6 +576,33 @@ TEST_F(GcsActorManagerTest, TestNamedActorDeletionNotHappendWhenReconstructed) {
             request1.task_spec().actor_creation_task_spec().actor_id());
 }
 
+TEST_F(GcsActorManagerTest, TestDestroyActorBeforeActorCreationCompleted) {
```
Comment (raulchen) [inline src/ray/gcs/gcs_server/test/gcs_actor_manager_test.cc:579]:
```suggestion
TEST_F(GcsActorManagerTest, TestDestroyActorBeforeActorCreationCompletes) {
```

File: src/ray/gcs/gcs_server/test/gcs_actor_manager_test.cc [hunk 2]
```diff
@@ -576,6 +576,33 @@ TEST_F(GcsActorManagerTest, TestNamedActorDeletionNotHappendWhenReconstructed) {
...
+  mock_actor_scheduler_->actors.clear();
+
+  ASSERT_TRUE(worker_client_->Reply());
+
+  // Check that the actor is in state `DEAD`.
+  rpc::Address address;
+  auto node_id = ClientID::FromRandom();
+  auto worker_id = WorkerID::FromRandom();
+  address.set_raylet_id(node_id.Binary());
+  address.set_worker_id(worker_id.Binary());
```
Comment (raulchen) [inline src/ray/gcs/gcs_server/test/gcs_actor_manager_test.cc:600]: Nit, put this in a helper function `RandomAddress`?

File: src/ray/gcs/gcs_server/test/gcs_actor_manager_test.cc [hunk 3]
```diff
@@ -576,6 +576,33 @@ TEST_F(GcsActorManagerTest, TestNamedActorDeletionNotHappendWhenReconstructed) {
...
+      create_actor_request, [&finished_actors](std::shared_ptr<gcs::GcsActor> actor) {
+        finished_actors.emplace_back(actor);
+      }));
+
+  ASSERT_EQ(finished_actors.size(), 0);
+  ASSERT_EQ(mock_actor_scheduler_->actors.size(), 1);
+  auto actor = mock_actor_scheduler_->actors.back();
+  mock_actor_scheduler_->actors.clear();
+
+  ASSERT_TRUE(worker_client_->Reply());
```
Comment (raulchen) [inline src/ray/gcs/gcs_server/test/gcs_actor_manager_test.cc:593]: please document why this is needed here.

File: src/ray/gcs/gcs_server/test/gcs_actor_manager_test.cc (hunk 1 above)
Comment (ffbin) [reply src/ray/gcs/gcs_server/test/gcs_actor_manager_test.cc]: fixed

File: src/ray/gcs/gcs_server/test/gcs_actor_manager_test.cc (hunk 2 above)
Comment (ffbin) [reply src/ray/gcs/gcs_server/test/gcs_actor_manager_test.cc]: fixed

File: src/ray/gcs/gcs_server/test/gcs_actor_manager_test.cc (hunk 3 above)
Comment (ffbin) [reply src/ray/gcs/gcs_server/test/gcs_actor_manager_test.cc:570]: fixed

**Review by raulchen — APPROVED** [review]

**Pull request merged**